memo/src/instruction.rs

```diff
@@ -0,0 +1,49 @@
+use crate::error::MemoError;
+use solana_sdk::{entrypoint::ProgramResult, info, program_error::ProgramError};
+use std::{mem::size_of, str::from_utf8};
+
+/// Instructions supported by the memo program
+#[repr(C)]
+#[derive(Clone, Debug, PartialEq)]
+pub enum Instruction<'a> {
+    /// Validate UTF-8 encoded characters
+    Utf8(&'a str),
```

Comment:
This'll cause a bincode prefix to the valid UTF-8 encoded characters right?  That's going to feel clunky for non-rust users.    The instruction data can just be the raw UTF-8

Reply:
Yes. This was @garious 's idea, with the thought that we might want to support different encoding types in future versions

Reply:
Right, thought we might want to add things like `AddEncryptedMemo` or `AddUtf16Memo` or `AddGifMemo`.  Just never know.  And currently we use Bincode everywhere.  If we're going to break away from Bincode, I'd like to make it for the purpose of supporting some language-agnostic RPC, like protobuf.  @jstarry is looking into it.

Reply:
I'm also okay with a Program ID per memo type, such that the `Instruction::data` can be the UTF-8 string with no header.  But so far, `Config` program is the only program that broke from the `enum Instruction` convention. Personally, I find that program to be one of the most confusing ones.

Reply:
That does sound appealing though:

```rust
let memo_instruction = Instruction {
    program_id: "Memo1UhkJRfHyvLMcVucJwxXeuD728EqVDDwQDxFMNo".parse().unwrap(),
    accounts: vec![],
    data = b"Hello, world",
};
```

Wouldn't even need a custom error, could just use `InstructionError::InvalidInstructionData`.

@CriesofCarrots, what do you think about throwing together a second PR with this approach so that we can compare and contrast?

Reply:
Sure, I can do that. To be clear, then, each encoding/content type would need a separate program, so this becomes the `utf8-memo` program or some such?

Reply:
or just keep it `memo` and force the next one (if it ever exists) to add precision. fine to let simple things be simple